docs/design/generic-enhanced-purls.md

```diff
@@ -0,0 +1,150 @@
+
+## Enhanced purls overview
+
+- Implement an `enhancer` for every supported purl type, which is essentially a set of rules that will be applied to a generic artifact and, in case those rules can be matched, it will replace the generic purl for a more specific type.
+    - The most generic rule seems to be the package's origin. With famous and well established public registries such as `maven.org` or `huggingface.co`, it is easy to correlate the package's origin with a more specific type.
+    - Mapping famous and public registries will prevent the use of private registries. To solve this, we need to allow users to change Cachi2's default configuration to enable additional source URLs for each purl type.
+    - More rules might be useful or even necessary when generating specific purl types. For example, Maven has the concept of groupId, artifactId and version (GAV) for each of its artifacts, and these values are fundamental in defining both the download URL and the purl.
+- Extend the generic artifacts lockfile specification to introduce a `type` attribute that allows users to hint at which purl type that artifact should have.
+    - The only types that can be selected are the ones that had their respective `enhancers` implemented. We don't want to allow users to be able to freely specify the purl type for a generic artifact, but rather restrict this use to a specific subset of purl types.
+    - Any failures to match the hinted `type` will not cause the request to fail. The file will still be fetched, but the purl will revert to `pkg:generic`.
+
+## A practical example
+
+### Input files
+
+**generic_artifacts.yaml**
+```yaml
+metadata:
+    version: '1.0'
+artifacts:
+  - download_url: https://github.com/containerbuildsystem/cachi2/archive/refs/tags/0.11.0.tar.gz
+    target: cachi2_0_11_0.tar.gz
+    checksums: 
+      sha256: fa0d536389db15fb3dabdb3b3d08354f47f765a653178140bfbe1b3de1a6ee76
+  - download_url: https://maven.repository.internal.com/ga/io/quarkus/quarkus-core/3.8.5.internal-00004/quarkus-core-3.8.5.internal-00004.jar
+    target: quakus.jar
+    type: maven
+    checksums:
+      sha1: e4ca5fadf89e62fb29d0d008046489b2305295bf
+  - download_url: https://huggingface.co/nvidia/Llama-3.1-Nemotron-70B-Instruct-HF/blob/b919e5d07ce15f31ea741f2be99a00a33c3b427b/model-00001-of-00030.safetensors
+    target: llama_3.1_1_of_30.safetensors
+    type: huggingface
+```
+
+**.cachi2-config.yaml**
+```yaml
+generic-artifact-sources:
+    maven:
+        # sample internal registry
+        - maven.repository.internal.com
+        - maven.org
+    huggingface:
+        - huggingface.co
+```
+
+### Cachi2 CLI usage
+
+```
+cachi2 fetch-deps --source /path/to/repo generic
```

Comment:
I think we'll just use/enhance the existing [config-file functionality](https://github.com/containerbuildsystem/cachi2?tab=readme-ov-file#configuration), right?

Reply:
Yes, we can default to a location within a repo and use defaut configuration mechanism, but would need a way to point to some external config which will be version-controlled elsewhere. Suppose someone needs to build ScaPy (or some other kilostarred package) and suppose they _also_ need some prebuilt generic artifact. I don't think it is safe to assume that upstream ScaPy would just accept any cachi2 yaml. If it is a local clone the someone would still need to add the config and maintain it. Likely not a blocker right now, but I feel like we'll need to deal with this at some point.

Reply:
I agree with some of the @a-ovchinnikov's reasoning when it comes to checking-in some random file, but then again, imagine a blackbox pipeline, where would such a file come from? It might easily become a devops engineer's nightmare. I also agree with @brunoapimentel  that from reproducibility POV and ease of debugging you want users to commit the file to their repos. Now, why would any random big upstream care about cachi2? Exactly, they won't, that's why downstream exists and then such a file is perfectly fine to be checked in to a repo. However, downstream isn't very popular in K8S world and so users tend to solve it with git submodules instead to compensate for the need of downstream control and here again, checking in the cachi2 lockfile is perfectly fine. I don't even think we need a different way of providing it ATM. If the need arises, we can discuss again.